// web-monetization-ad-hider/dist/wm-ad-hider.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.WebMonetizationAdHider = factory());
}(this, (function () { 'use strict';

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function");
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        writable: true,
        configurable: true
      }
    });
    if (superClass) _setPrototypeOf(subClass, superClass);
  }

  function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
      return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
  }

  function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
      o.__proto__ = p;
      return o;
    };

    return _setPrototypeOf(o, p);
  }

  function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  function _construct(Parent, args, Class) {
    if (_isNativeReflectConstruct()) {
      _construct = Reflect.construct;
    } else {
      _construct = function _construct(Parent, args, Class) {
        var a = [null];
        a.push.apply(a, args);
        var Constructor = Function.bind.apply(Parent, a);
        var instance = new Constructor();
        if (Class) _setPrototypeOf(instance, Class.prototype);
        return instance;
      };
    }

    return _construct.apply(null, arguments);
  }

  function _isNativeFunction(fn) {
    return Function.toString.call(fn).indexOf("[native code]") !== -1;
  }

  function _wrapNativeSuper(Class) {
    var _cache = typeof Map === "function" ? new Map() : undefined;

    _wrapNativeSuper = function _wrapNativeSuper(Class) {
      if (Class === null || !_isNativeFunction(Class)) return Class;

      if (typeof Class !== "function") {
        throw new TypeError("Super expression must either be null or a function");
      }

      if (typeof _cache !== "undefined") {
        if (_cache.has(Class)) return _cache.get(Class);

        _cache.set(Class, Wrapper);
      }

      function Wrapper() {
        return _construct(Class, arguments, _getPrototypeOf(this).constructor);
      }

      Wrapper.prototype = Object.create(Class.prototype, {
        constructor: {
          value: Wrapper,
          enumerable: false,
          writable: true,
          configurable: true
        }
      });
      return _setPrototypeOf(Wrapper, Class);
    };

    return _wrapNativeSuper(Class);
  }

  function _assertThisInitialized(self) {
    if (self === void 0) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return self;
  }

  function _possibleConstructorReturn(self, call) {
    if (call && (typeof call === "object" || typeof call === "function")) {
      return call;
    }

    return _assertThisInitialized(self);
  }

  function _createSuper(Derived) {
    var hasNativeReflectConstruct = _isNativeReflectConstruct();

    return function () {
      var Super = _getPrototypeOf(Derived),
          result;

      if (hasNativeReflectConstruct) {
        var NewTarget = _getPrototypeOf(this).constructor;

        result = Reflect.construct(Super, arguments, NewTarget);
      } else {
        result = Super.apply(this, arguments);
      }

      return _possibleConstructorReturn(this, result);
    };
  }

  var WebMonetizationAdHider = /*#__PURE__*/function (_HTMLElement) {
    _inherits(WebMonetizationAdHider, _HTMLElement);

    var _super = _createSuper(WebMonetizationAdHider);

    function WebMonetizationAdHider() {
      var _this;

      _classCallCheck(this, WebMonetizationAdHider);

      _this = _super.call(this);
      _this.hasPaid = false;
      _this.templateSelector = _this.getAttributeOrFallback("template", null);
      _this.timeout = parseInt(_this.getAttributeOrFallback("timeout", "3000"), 10);
      return _this;
    }

    _createClass(WebMonetizationAdHider, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this2 = this;

        if (this.templateSelector === null) {
          this.adCode = this.querySelector("template");
        } else {
          this.adCode = document.querySelector(this.templateSelector);
        }

        if (document.monetization && this.adCode) {
          this.hasPaid = document.monetization.state === "started";

          if (!this.hasPaid) {
            this.monetizationListener = function () {
              _this2.hasPaid = true;

              _this2.removeAds();
            };

            document.monetization.addEventListener("monetizationstart", this.monetizationListener);
            window.setTimeout(function () {
              if (!_this2.hasPaid) {
                _this2.showAds();
              }
            }, this.timeout);
          }
        } else {
          this.showAds();
        }
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (document.monetization && this.monetizationListener) {
          document.monetization.removeEventListener("monetizationstart", this.monetizationListener);
        }
      }
    }, {
      key: "showAds",
      value: function showAds() {
        if (this.adCode && this.adCode.content) {
          var ads = this.adCode.content.cloneNode(true);
          this.appendChild(ads);
        }
      }
    }, {
      key: "removeAds",
      value: function removeAds() {
        while (this.firstChild) {
          this.removeChild(this.firstChild);
        }
      }
    }, {
      key: "getAttributeOrFallback",
      value: function getAttributeOrFallback(attributeName, fallback) {
        if (this.hasAttribute(attributeName)) {
          return this.getAttribute(attributeName);
        }

        return fallback;
      }
    }]);

    return WebMonetizationAdHider;
  }( /*#__PURE__*/_wrapNativeSuper(HTMLElement));

  if ("customElements" in window) {
    customElements.define("wm-ad-hider", WebMonetizationAdHider);
  }

  return WebMonetizationAdHider;

})));
